api/v1beta1/azuremanagedcontrolplanetemplate_validation.go

```diff
@@ -0,0 +1,273 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"reflect"
+	"strings"
+
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Validate the Azure Managed Control Plane Template and return an aggregate error.
+func (mcp *AzureManagedControlPlaneTemplate) validateManagedControlPlaneTemplate(cli client.Client) error {
+	validators := []func(client client.Client) error{
+		mcp.validateName,
+		mcp.validateVersion,
+		mcp.validateDNSServiceIP,
+		mcp.validateLoadBalancerProfile,
+		mcp.validateManagedClusterNetwork,
+		// mcp.validateAutoScalerProfile,
+	}
+
+	var errs []error
+	for _, validator := range validators {
+		if err := validator(cli); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return kerrors.NewAggregate(errs)
+}
+
+// validateDNSServiceIP validates the DNSServiceIP.
+func (mcp *AzureManagedControlPlaneTemplate) validateDNSServiceIP(_ client.Client) error {
```

Comment:
@CecileRobertMichon A quick question regarding repeated code. A lot of webhook code is being repeated from [azuremanagedcontrolplane_webhook.go](https://github.com/willie-yao/cluster-api-provider-azure/blob/ee91ad8bb881748b7a24796965ce94b8537fe4af/api/v1beta1/azuremanagedcontrolplane_webhook.go#L282). Would it be better design-wise to refactor the functions used to take in the template fields as parameters so that the functions can be re-used? I observed a similar pattern in [azurecluster_validation.go](https://github.com/willie-yao/cluster-api-provider-azure/blob/ee91ad8bb881748b7a24796965ce94b8537fe4af/api/v1beta1/azurecluster_validation.go#L271).

Reply:
Update: I refactored the re-used functions to be usable for both `AzureManagedControlPlanes` and `AzureManagedControlPlaneTemplates`.